### PR TITLE
--standalone type forwarding

### PR DESCRIPTION
**Pull request by KevinRansom**
There has long been a bug with the coreclr version of the compiler:

**Fixes: #3924 : Support for --standalone on .NET Core**

This happens because --standalone takes the Il from FSharp.Core and embeds it in the binary being built.  When FSharp.Core references framework assemblies and those assemblies rely on type forwarders to get to the correct type the compiler does not know how to unify the generated method signatures with the type forwarded ones.

The fix in this PR is to follow the type forwarders for everything in thethe generated code and embedded code.
_This only happens in the static linker, and so will have no impact to the majority of applications._

**Static linking occurs for:**
1. Apps compiled with --standalone
2. Apps that rely on generative type providers
3. Apps that use --staticlink: to link multiple assemblies into a single dll

This fix is also necessary for us to move to a single netstandard 2.0 version of FSharp.Core rather than dual desktop and netstandard versions.  Because even the desktop --standalone which has always work would start failing for that scenario.

Note:
         The pr also contains this fix: https://github.com/dotnet/fsharp/pull/7461

### CODE REVIEW

File: src/fsharp/fsc.fs [hunk 1]
```diff
@@ -1201,11 +1498,11 @@ module StaticLinker =
...
                     mkILFields (topTypeDefs |> List.collect (fun td -> td.Fields.AsList)))
 
-            let ilxMainModule = 
-                { ilxMainModule with 
+            let ilxMainModule =
+                { ilxMainModule with
                     Manifest = (let m = ilxMainModule.ManifestOfAssembly in Some {m with CustomAttrsStored = storeILCustomAttrs (mkILCustomAttrs (m.CustomAttrs.AsList @ savedManifestAttrs)) })
                     CustomAttrsStored = storeILCustomAttrs (mkILCustomAttrs [ for m in moduls do yield! m.CustomAttrs.AsArray ])
-                    TypeDefs = mkILTypeDefs (topTypeDef :: List.concat normalTypeDefs)
+                    TypeDefs = typeForwarding.TypeForwardILTypeDefs(topTypeDef :: List.concat normalTypeDefs)
```
Comment (dsyme) [inline src/fsharp/fsc.fs]: Can the 400 lines of code from line 1175 be replaced by 

    let ilxMainModule2 = morphILTypeRefsInILModuleMemoized typeForwardILTypeRef ilxMainModule 

thanks

Comment (KevinRansom) [reply src/fsharp/fsc.fs]: @dsyme, done … I wish I had discovered that function a week ago.

File: src/fsharp/fsc.fs [hunk 2]
```diff
@@ -1109,13 +1109,80 @@ module MainModuleBuilder =
...
+                                | None -> scoref
+                            result
+                        | None -> scoref
+                    | false, _ -> scoref
+            | _ -> scoref
+
+        let typeForwardILTypeRef (tref: ILTypeRef) =
+            let scoref1 = tref.Scope
+            let scoref2 = followTypeForwardForILTypeRef tref
+            if scoref1 === scoref2 then tref
```
Comment (cartermp) [inline src/fsharp/fsc.fs]: > `===`

wat

Comment (KevinRansom) [reply src/fsharp/fsc.fs]: :-)
I read that operator as "is the same as"
It is defined here:   https://github.com/dotnet/fsharp/blob/master/src/absil/illib.fs#L44

Comment (cartermp) [reply src/fsharp/fsc.fs]: ah, so it's `objEq (box x) (box y)`

Comment (cartermp) [reply src/fsharp/fsc.fs]: why not `=`?

Comment (KevinRansom) [reply src/fsharp/fsc.fs]: It's not instance equality, it's are the contents equal.

**Pull request merged**